src/sources/file/line_agg.rs

```diff
@@ -0,0 +1,519 @@
+use bytes::{Bytes, BytesMut};
+use futures::{Async, Poll, Stream};
+use regex::bytes::Regex;
+use serde::{Deserialize, Serialize};
+use std::collections::{HashMap, VecDeque};
+use std::time::Duration;
+use tokio::timer::DelayQueue;
+
+#[derive(Debug, Hash, Clone, PartialEq, Deserialize, Serialize)]
+#[serde(rename_all = "snake_case")]
+pub enum Mode {
+    /// All consecutive lines matching this pattern are included in the group.
+    /// The first line (the line that matched the start pattern) does not need
+    /// to match the `ContinueThrough` pattern.
+    /// This is useful in cases such as a Java stack trace, where some indicator
+    /// in the line (such as leading whitespace) indicates that it is an
+    /// extension of the preceeding line.
+    ContinueThrough,
+
+    /// All consecutive lines matching this pattern, plus one additional line,
+    /// are included in the group.
+    /// This is useful in cases where a log message ends with a continuation
+    /// marker, such as a backslash, indicating that the following line is part
+    /// of the same message.
+    ContinuePast,
+
+    /// All consecutive lines not matching this pattern are included in the
+    /// group.
+    /// This is useful where a log line contains a marker indicating that it
+    /// begins a new message.
+    HaltBefore,
+
+    /// All consecutive lines, up to and including the first line matching this
+    /// pattern, are included in the group.
+    /// This is useful where a log line ends with a termination marker, such as
+    /// a semicolon.
+    HaltWith,
+}
```

Comment:
Thanks! This kind of documentation is excellent! I would only recommend putting a few words before these new types in case we run `cargo doc` on Vector some day.

Reply:
Good point!
I really want `#[deny(missing_docs)]`. Maybe I’ll enable it for just this mod, as an experiment.